eth/downloader/queue.go

```diff
@@ -49,35 +56,73 @@ type fetchRequest struct {
+	fastSyncPivot uint64   // Block number where the fast sync pivots into archive synchronisation mode
+
 	hashPool    map[common.Hash]int // [eth/61] Pending hashes, mapping to their insertion index (priority)
 	hashQueue   *prque.Prque        // [eth/61] Priority queue of the block hashes to fetch
 	hashCounter int                 // [eth/61] Counter indexing the added hashes to ensure retrieval order
 
-	headerPool  map[common.Hash]*types.Header // [eth/62] Pending headers, mapping from their hashes
-	headerQueue *prque.Prque                  // [eth/62] Priority queue of the headers to fetch the bodies for
-	headerHead  common.Hash                   // [eth/62] Hash of the last queued header to verify order
+	headerHead common.Hash // [eth/62] Hash of the last queued header to verify order
+
+	blockTaskPool  map[common.Hash]*types.Header // [eth/62] Pending block (body) retrieval tasks, mapping hashes to headers
+	blockTaskQueue *prque.Prque                  // [eth/62] Priority queue of the headers to fetch the blocks (bodies) for
+	blockPendPool  map[string]*fetchRequest      // [eth/62] Currently pending block (body) retrieval operations
+	blockDonePool  map[common.Hash]struct{}      // [eth/62] Set of the completed block (body) fetches
+
+	receiptTaskPool  map[common.Hash]*types.Header // [eth/63] Pending receipt retrieval tasks, mapping hashes to headers
+	receiptTaskQueue *prque.Prque                  // [eth/63] Priority queue of the headers to fetch the receipts for
+	receiptPendPool  map[string]*fetchRequest      // [eth/63] Currently pending receipt retrieval operations
+	receiptDonePool  map[common.Hash]struct{}      // [eth/63] Set of the completed receipt fetches
 
-	pendPool map[string]*fetchRequest // Currently pending block retrieval operations
+	stateTaskIndex int                      // [eth/63] Counter indexing the added hashes to ensure prioritized retrieval order
+	stateTaskPool  map[common.Hash]int      // [eth/63] Pending node data retrieval tasks, mapping to their priority
+	stateTaskQueue *prque.Prque             // [eth/63] Priority queue of the hashes to fetch the node data for
+	statePendPool  map[string]*fetchRequest // [eth/63] Currently pending node data retrieval operations
 
-	blockPool   map[common.Hash]uint64 // Hash-set of the downloaded data blocks, mapping to cache indexes
-	blockCache  []*Block               // Downloaded but not yet delivered blocks
-	blockOffset uint64                 // Offset of the first cached block in the block-chain
+	stateDatabase   ethdb.Database   // [eth/63] Trie database to populate during state reassembly
+	stateScheduler  *state.StateSync // [eth/63] State trie synchronisation scheduler and integrator
+	stateProcessors int32            // [eth/63] Number of currently running state processors
+	stateSchedLock  sync.RWMutex     // [eth/63] Lock serializing access to the state scheduler
```

Comment:
1. Done, will push in next commit.
2. Actually, the code is split between Mu and Lock. If not for other reason, because I personally used Lock everywhere :stuck_out_tongue_winking_eye: . On a more serious tone though, I quite heavily oppose to leaking type information into the variable name. Calling it a `Mu(tex)` is a tad much imho.